### ISSUE TITLE
ESO retrieve data - Files keep failing

### ISSUE THREAD

**Issue by jorgehumberto** [astropy/astroquery#2622]
HI,

I am attempting to download some ESPRESSO data from ESO's archive via the comment line using astroquery. Here's a MWE of what I do :

```
from astroquery.eso import EsoClass
filenames = ["ESPRE.2022-10-06T01:18:07.889"]
class esoArchiveConnector(EsoClass):
	
	def __init__(self) :
		super().__init__()
		
	def connect_to_eso_archive(self):
		
		
		self.ROW_LIMIT = -1
		# need to use special access link 
		self.QUERY_INSTRUMENT_URL = 'https://archive.eso.org/wdb/wdb/cas'
		self.USERNAME = my_username
		self.login(self.USERNAME, store_password=True)
		
		return

	def download_files(self,fileList:list):
		  
		for index, observation in enumerate(sorted(fileList)):
			print (observation)
			data_files = self.retrieve_data(observation, request_all_objects=False, unzip=False, continuation=True)
	 
	 
esoArchive = esoArchiveConnector()


esoArchive.connect_to_eso_archive()


esoArchive.download_files(filenames)
```

and I get the following errors:

```
Traceback (most recent call last):
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/urllib3/connection.py", line 174, in _new_conn
    conn = connection.create_connection(
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/urllib3/util/connection.py", line 96, in create_connection
    raise err
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/urllib3/util/connection.py", line 86, in create_connection
    sock.connect(sa)
TimeoutError: [Errno 110] Connection timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/urllib3/connectionpool.py", line 699, in urlopen
    httplib_response = self._make_request(
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/urllib3/connectionpool.py", line 394, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/urllib3/connection.py", line 239, in request
    super(HTTPConnection, self).request(method, url, body=body, headers=headers)
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/http/client.py", line 1256, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/http/client.py", line 1302, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/http/client.py", line 1251, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/http/client.py", line 1011, in _send_output
    self.send(msg)
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/http/client.py", line 951, in send
    self.connect()
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/urllib3/connection.py", line 205, in connect
    conn = self._new_conn()
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/urllib3/connection.py", line 186, in _new_conn
    raise NewConnectionError(
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7f7cd36202b0>: Failed to establish a new connection: [Errno 110] Connection timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/requests/adapters.py", line 489, in send
    resp = conn.urlopen(
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/urllib3/connectionpool.py", line 755, in urlopen
    retries = retries.increment(
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/urllib3/util/retry.py", line 574, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='dataportal.eso.org', port=10001): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f7cd36202b0>: Failed to establish a new connection: [Errno 110] Connection timed out'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test_eso_dl.py", line 34, in <module>
    esoArchive.download_files(filenames)
  File "test_eso_dl.py", line 25, in download_files
    data_files = self.retrieve_data(observation, request_all_objects=False, unzip=False, continuation=True)
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/astroquery/eso/core.py", line 853, in retrieve_data
    filename = self._request("GET", fileLink, save=True,
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/astroquery/query.py", line 296, in _request
    response = self._download_file(url, local_filepath, cache=cache,
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/astroquery/eso/core.py", line 613, in _download_file
    resp = super(EsoClass, self)._download_file(url, local_filepath,
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/astroquery/query.py", line 353, in _download_file
    response = self._session.request(method, url,
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/requests/sessions.py", line 587, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/requests/sessions.py", line 723, in send
    history = [resp for resp in gen]
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/requests/sessions.py", line 723, in <listcomp>
    history = [resp for resp in gen]
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/requests/sessions.py", line 266, in resolve_redirects
    resp = self.send(
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/requests/sessions.py", line 701, in send
    r = adapter.send(request, **kwargs)
  File "/home/espresso/.conda/envs/exo2_tools/lib/python3.8/site-packages/requests/adapters.py", line 565, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='dataportal.eso.org', port=10001): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f7cd36202b0>: Failed to establish a new connection: [Errno 110] Connection timed out'))
```

Any idea on what is going on?

Some comments:
- I only present one observation ID, but it fails with all
- a similar code used to work a few months ago, with no issue, now it does not
- I also tested with the 'esocas' approach described in issue 1360 (https://github.com/astropy/astroquery/issues/1360)
- Using the link generated inside 'astroquery' I can download the file on my browser
- my credentials are working fine
- I installed/uninstalled/reinstalled the latest stable versions of 'astroquery' and requests

**Comment by keflavich** [comment]
This looks like an upstream issue, given that all of your errors are timeout errors.  Have you tried via the ESO archive website?

**Comment by jorgehumberto** [comment]
Hi Adam,

Yes,I have, and I can download the file with no issues. 

Cheers
Jorge

⁣

**Comment by Pharisaeus** [comment]
The immediate issue is:

> host='dataportal.eso.org', port=10001

There is no such thing, so a connection request for it will always fail. I am curious where this port number comes from. It might suggests some reverse-proxy misconfiguration, because I don't think astroquery generates links like that, so it must be some weird redirect.

In the meantime, if you just want to download files by id (without expansion or calselector), you can simply do it via:

https://dataportal.eso.org/dataPortal/file/FILE_ID

and if the file is proprietary, you can use the same link, but include credentials with the request (can be basic, oidc token or even session cookie), so the most trivial approach would be

```python
requests.get("https://dataportal.eso.org/dataPortal/file/FILE_ID", headers={"Authorization": b"Basic " + base64.b64encode(b'USER:PASSWORD')})
```

Hard to say something more, because I don't really understand the astroquery code, and it's doing some crazy stuff with old ESO Data Portal "API".

**Comment by jorgehumberto** [comment]
@Pharisaeus 

Thank you for the suggestion, it would have been perfect for what I want. 

However when I use that link it always fails (even when I copy/paste the link in the browser):
e.g.: 
```
    https://dataportal.eso.org/dataPortal/file/ESPRE.2022-10-06T01:18:07.889
```

I also tried:
```
    https://dataportal.eso.org/dataPortal/file/ESPRE.2022-10-06T01:18:07.889.fits
    https://dataportal.eso.org/dataPortal/file/ESPRE.2022-10-06T01:18:07.889.fits.Z
```

Could it be because I need to use the special access server?

Cheers
Jorge

**Comment by Pharisaeus** [comment]
@jorgehumberto 401 means you don't have access to the file :) If you just copypaste link in the browser then you're most likely not sending any authentication credentials, so the access to proprietary files (which is the case, because 2022-10 is not yet a year old) is denied.

With python requests it should be ok as long as you're sending valid user credentials with access to that file. So if you get 401 it means that particular user does not have access to file in question (eg. is not a PI). Are you sure you're using the right credentials?

One way to make sure you didn't misspell something is to request and pass around a token:

```python
def get_token(username, password):
    token_url = "https://www.eso.org/sso/oidc/oidcAccessToken"
    response = requests.get(token_url, params={"grant_type": "password", "client_id": "clientid", "username": username, "password": password})
    return json.loads(response.content)['id_token']

token = get_token(user, password)
r = requests.get("https://dataportal.eso.org/dataPortal/file/FILE_ID", headers={"Authorization": b"Bearer " + token})
```

This way if credentials are wrong the code will crash while trying to get a token.

**Comment by jorgehumberto** [comment]
Hi @Pharisaeus 

Yes, it looks I spoke too soon. I manage to make a request to check the available data, so I assumed I was login in correclty, but that is not the case. As you say, the authentication credentials are now being sent. I'll try your tokenb approach and get back to you as soon as possible.

Cheers
Jorge

**Comment by jorgehumberto** [comment]
Hi @Pharisaeus 

Ok, I've tested your code and I am still not able to get the data programatically. However, I can dowload the generated link "https://dataportal.eso.org/dataportal/file/ESPRE.2022-10-06T01:18:07.889" with no issues on the browser. To be sure I don't mess up, I pass the credentials to the token function as two strings, which I copy/pasted directly from my password manager to avoid errors. 

The only thjng I had to modify on your code was `b"Bearer " + token` to `"Bearer " + token` (or even `bytes(f"Bearer {token}", 'ascii')` or `bytes(f"Bearer {token}", 'utf-8)`) as it would fail trying to concatenate bytes and string objects.

So.. my guess is that I cannot pass the token or credentials to the request. I even tested some of the codes at http://archive.eso.org/programmatic/#SCRIPT and get the same issue.

Could it be a version issue? I am using 

```
>>> python_version()
'3.8.12'
>>> requests.__version__
'2.28.1'
```
and
```
CentOS Linux release 7.9.2009 (Core)
```

Any idea?

Cheers
Jorge

**Comment by Pharisaeus** [comment]
@jorgehumberto

> However, I can dowload the generated link "https://dataportal.eso.org/dataportal/file/ESPRE.2022-10-06T01:18:07.889" with no issues on the browser.

That's due to a small "trick"  we do when you enter the download portal page and log-in -> it propagates a cookie between archive.eso.org and dataportal.eso.org, so for some time the links will work, even for proprietary files, because the browser will send the cookie.

If the `get_token()` function does not crash and actually returns something, then the credentials have to be correct.
So what exactly happens when you do:

```python
r = requests.get("https://dataportal.eso.org/dataPortal/file/ESPRE.2022-10-06T01:18:07.889", headers={"Authorization": "Bearer " + token})
```
You still get 401?

**Comment by jorgehumberto** [comment]
Yes, I still get 401.

Not sure if it helps, but when I print the headers (`print(r.request.headers )`) I get:

```
{'User-Agent': 'python-requests/2.28.1', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'Connection': 'keep-alive', 'Authorization': 'Basic TlNBTlRPUzpudW5vZXNvMTIzNA=='}
```

Shouldn't the "Authorization" be "Bearer TOKEN"

and yes, I'am 100% sure the user I'm using has acces to the file, to be sure I copy/pasted the username and pass from my password manager (which I use for the browser access)

**Comment by Pharisaeus** [comment]
@jorgehumberto please edit the above comment because Basic auth string contains essentially plaintext username and password!

Anyway, the issue is that your user `NS***` does not have access to said file :( I've seen in the logs also other requests for that file, from a different user `esp***` who does have access.

**Comment by jorgehumberto** [comment]
OK, then we're getting somewhere, it should be using the `esp***`user (which is the one I specify in my code) and not the `NS***` one...

**Comment by Pharisaeus** [comment]
> Shouldn't the "Authorization" be "Bearer MY_TOKEN"?

Indeed, are you sure you're running the code you think you're running? :) Maybe you edited a different script and you're running a different one?

**Comment by jorgehumberto** [comment]
ehehe, I am , I have the bad habit of adding random prints when I make minor changes ;)

now, I'd like to understand why is my code is sending the requests as `NS***`. This user is from my institute, but should send their requests from their laptop and not our server.

Could it be that we share the same IP? does not make any sense to me...

**Comment by Pharisaeus** [comment]
Definitely nothing to do with IP, notice that the code you showed before `print(r.request.headers )` somehow magically included the basic auth header for that particular user. So the `request` you sent contained those credentials somehow! The issue is somewhere on your side, hard to say where exactly.

One crazy idea I can think of is if said user accidentally modified the requests library on the machine you're using and hardcoded their credentials to be automatically included in every request, but that would be very weird.

**Comment by jorgehumberto** [comment]
that would be really weird indeed, specially because said user does not have access to the env I am using. besides I removed and re-installed requests just in case (speccaily because I had added a ton of test prints in the mid of it 😅😅😅)

thanks for all the help, at least I "know" where to point my guns at... or at least that the problem is on my side, albeit no idea where... 😅😅😅

**Comment by Pharisaeus** [comment]
I'm not sure if the two things are related, but if your python-requests are adding random unexpected headers then who knows what else might be happening and what impact it has...

**Comment by jorgehumberto** [comment]
Found the problem and solution thanks to this:
https://stackoverflow.com/a/35378429
the credentials from the other user were hardcodded into ~/.netrc

all working fine now, managed to download the file sucessfully, checked the header and all good. Now I need to find a way to make requests ignore that file. And now I remember that for a couple of weeks at the begginging of the GTO, we used the `NS***
 user until the PI gave acess to the `esp***` user...

thanks for the help, would have never been able to figure this out without it!
